File: ioi/ioi01p1/ioi01p1.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.Reader;

class BIT2D {
	private int[][] data;
	private int gx, gy;

	BIT2D(int gx, int gy) {
		data = new int[gy + 1][gx + 1];
		this.gx = gx;
		this.gy = gy;
	}

	void add(int x, int y, int delta) {
		for (; y <= gy; y += -y & y) {
			for (int x1 = x; x1 <= gx; x1 += -x1 & x1) {
				data[y][x1] += delta;
			}
		}
	}

	private int query(int x, int y) {
		int ret = 0;
		for (; y > 0; y -= -y & y) {
			for (int x1 = x; x1 > 0; x1 -= -x1 & x1) {
				ret += data[y][x1];
			}
		}
		return ret;
	}

	int sum(int x1, int y1, int x2, int y2) {
		return query(x2, y2) + query(x1 - 1, y1 - 1) - query(x2, y1 - 1) - query(x1 - 1, y2);
	}
}

class FastReader extends BufferedReader {
	FastReader(Reader in) {
		super(in);
	}

	char nextChar() throws Exception {
		int c;
		while ((c = read()) < 33);
		return (char) c;
	}

	double nextDouble() throws Exception {
		int c;
		double div = 1;
		double ret = 0;
		while ((c = read()) < 33);
		boolean neg = c == 45;
		if (neg) {
			c = read();
		}
		do {
			ret = ret * 10 + c - 48;
		} while ((c = read()) > 47 && c < 58);
		if (c == 46) {
			while ((c = read()) > 47 && c < 58) {
				ret += (c - 48) / (div *= 10);
			}
		}
		return neg ? -ret : ret;
	}

	int nextInt() throws Exception {
		int c;
		int ret = 0;
		while ((c = read()) < 33);
		boolean neg = c == 45;
		if (neg) {
			c = read();
		}
		do {
			ret = ret * 10 + c - 48;
		} while ((c = read()) > 47 && c < 58);
		return neg ? -ret : ret;
	}

	long nextLong() throws Exception {
		int c;
		long ret = 0;
		while ((c = read()) < 33);
		boolean neg = c == 45;
		if (neg) {
			c = read();
		}
		do {
			ret = ret * 10 + c - 48;
		} while ((c = read()) > 47 && c < 58);
		return neg ? -ret : ret;
	}
}

public class ioi01p1 {
	private static FastReader in;
	private static PrintWriter out;

	public static void main(String[] args) throws Exception {
		in = new FastReader(new InputStreamReader(System.in));
		out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
		in.nextInt();
		int s = in.nextInt();
		BIT2D bit = new BIT2D(s, s);
		int i;
		do {
			i = in.nextInt();
			if (i == 1) {
				int x = in.nextInt();
				int y = in.nextInt();
				int a = in.nextInt();
				bit.add(x + 1, y + 1, a);
			} else if (i == 2) {
				int l = in.nextInt();
				int b = in.nextInt();
				int r = in.nextInt();
				int t = in.nextInt();
				out.println(bit.sum(l + 1, b + 1, r + 1, t + 1));
			}
		} while (i < 3);
		in.close();
		out.close();
	}
}

File: ioi/ioi01p1/ioi01p1.py
class BIT2D:
	def __init__(self, gx, gy):
		self.data = [[0] * (gx + 1) for _ in xrange(gy + 1)]
		self.gx = gx
		self.gy = gy

	def add(self, x, y, delta):
		while y <= self.gy:
			x1 = x
			while x1 <= self.gx:
				self.data[y][x1] += delta
				x1 += -x1 & x1
			y += -y & y

	def query(self, x, y):
		ret = 0
		while y:
			x1 = x
			while x1:
				ret += self.data[y][x1]
				x1 -= -x1 & x1
			y -= -y & y
		return ret

	def sum(self, x1, y1, x2, y2):
		return self.query(x2, y2) + self.query(x1 - 1, y1 - 1) - self.query(x2, y1 - 1) - self.query(x1 - 1, y2)


input = __import__("sys").stdin.readline
s = int(input().split()[1])
bit = BIT2D(s, s)
data = map(int, input().split())
while data[0] < 3:
	if data[0] == 1:
		x, y, a = data[1:]
		bit.add(x + 1, y + 1, a)
	elif data[0] == 2:
		l, b, r, t = data[1:]
		print(bit.sum(l + 1, b + 1, r + 1, t + 1))
	data = map(int, input().split())